posts/2018-08-25-http4s-error-handling-mtl.md

```diff
@@ -0,0 +1,327 @@
+
+### Http Error Handling
+
+Okay let's just go ahead and add some error handling to our http route by taking advantange of the `MonadError` instance defined by our constraint `Sync[F]` and making use of the syntax provided by `cats`:
+
+```tut:book:silent
+class UserRoutesAlt[F[_]: Sync](userAlgebra: UserAlgebra[F]) extends Http4sDsl[F] {
+
+  import JsonCodecs._
+
+  val routes: HttpRoutes[F] = HttpRoutes.of[F] {
+
+    case GET -> Root / "users" / username =>
+      userAlgebra.find(username).flatMap {
+        case Some(user) => Ok(user)
+        case None => NotFound(username)
+      }
+
+    case req @ POST -> Root / "users" =>
+      req.as[User].flatMap { user =>
+        userAlgebra.save(user) *> Created(user.username)
+      }.handleErrorWith { // compiles without giving you "match non-exhaustive" error
+        case UserAlreadyExists(username) => Conflict(username)
+      }
+
+    case req @ PUT -> Root / "users" / username =>
+      req.as[UserUpdateAge].flatMap { userUpdate =>
+        userAlgebra.updateAge(username, userUpdate.age) *> Ok(username)
+      }.handleErrorWith { // compiles without giving you "match non-exhaustive" error
+        case InvalidUserAge(age) => BadRequest(s"Invalid age $age")
+      }
+  }
+
+}
+```
+
+Now we can say this implementation is quite elegant! We are handling and mapping business errors to the according http response and our code compiles without any warning whatsoever. But wait... We are not handling the `UserNotFound` error and the compiler didn't tell us about it! That's not cool and we as functional programmers believe in types because we can know what a function might do just by looking at the types but here it seems we hit the wall.
+
+The problem is that our constraint of type `Sync` from `cats-effect` has a `MonadError` instance with its type error fixed as `Throwable`. So the compiler can't help us here since this type is too generic. And we can't add a constraint for `MonadError[F, UserError]` because we would get an "ambigous implicits" error with two instances of `MonadError` in scope.
+
+So, what can we do about it?
+
+### Next level MTL: Optics
+
+I heard sometime ago about `Classy Optics` (Lenses, Prisms, etc) when I was learning `Haskell` and watched [this amazing talk](https://www.youtube.com/watch?v=GZPup5Iuaqw) by George Wilson but I never got to use this concept in Scala until now!
+
+Well first, let me give you a quick definition of `Lenses` and `Prisms`. In a few words we can define:
+
+- `Lenses` as getters and setters that compose making the accessing of nested data structure's fields quite easy.
+- `Prisms` as first-class pattern matching that let us access branches of an ADT and that also compose.
+
+And the idea of `Classy Optics` is to "associate each type a typeclass full of optics for that type".
```

Comment:
I think there's a missing "with" here.  And "associating each type with a typeclass" sounds funny to me.  Maybe "provide instances of a typeclass"?

Reply:
I literally tried to take George's definition from [his slides](https://github.com/gwils/next-level-mtl-with-classy-optics/blob/master/Slides.pdf) but I even failed doing that. Hope the correction is fine!